3.workload.md

```diff
@@ -0,0 +1,378 @@
+# 3. Workload
+
+This section defines the workload definition schematics.
+
+Workload definitions are provided by the platform (just as traits are) so that users may inspect the platform and
+ learn what workloads are available for use. Workload definitions are not extensible to platform users (only to platform
+  operators). Thus, platform users MUST NOT be allowed to create new workload kinds.
+
+## Top-Level Attributes
+
+The following attributes are common across all schemata defined in this document. This structure is designed to allow
+ object type to be determined by scanning two fields (`apiVersion` and `kind`), and provide access to common fields
+  in the `metadata` section. The `spec` section is where type-specific information is located.
+
+Here are the attributes that provide top-level information about the workload definition.
+
+| Attribute | Type | Required | Default Value | Description |
+|-----------|------|----------|---------------|-------------|
+| `apiVersion` | `string` | Y | | The specific version of the specification in use. The core types use `core.oam.dev/v1alpha2` |
+| `kind` | `string` | Y || The string `WorkloadDefinition` |
+| `metadata` | [`Metadata`](#metadata) | Y | | WorkloadDefinition metadata. |
+| `spec`| [`Spec`](#spec) | Y | | A container for the workload definition. |
+
+### Metadata
+
+The metadata describes this workload definition. The `metadata` section is defined in [the meta section of part 2](2.overview_and_terminology.md#Metadata)
```

Comment:
I also hope we can directly reference `metadata` from Kubernetes resource model (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#objectmeta-v1-meta) in "overview and terminology" doc. 

I'm not in favor of making the spec tedious and complex due to explanations on resource model and metadata.

Reply:
Same as above. Btw, this section is copied from the existing spec

Reply:
This doc is better: https://github.com/kubernetes/community/blob/1d6d124/contributors/devel/sig-architecture/api-conventions.md#metadata

My opinion is we need to have a clearer relationship between overview_terminology.md section and K8s resource model than today. /cc @negz 

We can fix this in following up PR of course.

Reply:
I'm hesitant about linking directly to the Kubernetes docs for formal definitions, including structure or field names. If Kubernetes decides to suddenly change then name or usage of a field, we would need to first assess that change and decide if it should be applied here too. Conceptual docs are probably fine since a change to the wording of a concept in the Kubernetes doc won't necessarily break an OAM runtime.

Reply:
@vturecek Actually, `.../generated/kubernetes-api/{version}` is the permanent link of versioned K8s api spec. It's much safer than what OAM spec currently linked in the doc, e.g.: https://github.com/oam-dev/spec/blob/master/2.overview_and_terminology.md#label-format which only reflects the latest version of K8s api.

Also, I propose we directly refer K8s metadata instead of creating a "simplified" version, because in practice, the OAM K8s runtime still need to use full version of [k8s metadata](https://github.com/oam-dev/rudr/blob/319448fb4cfd970152ecb24d3722536d66ea6bd7/examples/components.yaml#L3). The OAM spec version of "name + labels + annotations" actually is not enough to work at all. In that sense, current rudr doesn't fully follow OAM spec.

I'm totally OK we fix it in following PR but we need to fix it ASAP indeed.

Reply:
What I proposed above also applies to GVK section and many places in "overview and terminology" doc . Non-k8s impl will also be benefit a lot since it could import K8s types directly in many languages. It doesn't make sense to create sth "similar" in OAM spec.

Reply:
So are we saying that OAM-compliant runtimes must implement the full Kubernetes metadata object model, as specified in the Kubernetes API conventions doc? That seems extreme to me. Namespaces would be required by every runtime even though that has never been a concept we've had in OAM and to me seems outside of the scope of an application modeling definition.

As I stated earlier, I'm fine linking off to _concepts_ where we feel they are identical. If the content is versioned, all the better. 

I liked the suggestion earlier that we clearly label parts of the spec with the role they are most relevant to (app developer, app operator, infra operator). But I'm not sure what it means to have a separate part of the spec for platform builders, or what we gain from that. If they are split then I'm assuming there are object models we define for platform builders that end users don't see, in which case I'm not sure why we even care about those.

Reply:
@vturecek I also agree directly referring Kubernetes design doc is a good start to make the spec doc simpler.

While after digesting your comment above, I feel that the OAM spec doc cares very much about end users. But honestly, my personal observation is the current spec is barely friendly to them. I'd guess it's also the main reason that OAM spec was commented as "complex":  it reads like a doc for someone to implement an OAM platform rather than material to grape OAM concepts immediately at first glance, or more possibly, those two are inextricably mixed together in the spec.

In that sense, it's exactly the same mistake that K8s Deployment is making and we claim to solve in OAM :-)

Another concern is, the current spec is not implementers friendly either. For example, from reading the spec, I have feeling that we are suggesting implementers to code OAM metadata types from scratch. I really doubt there's any implementer will do that. /cc @negz 

So I don't mind if it's named "splitting", but I'd highlight the idea like: OAM spec **should also consider separate concerns** in its content, specifically:

1. straightforward docs for end users to get the core ideas of OAM in 10 min instead of reading a 10 pages specification. IMO, detailed fields of OAM metadata is not part of this. 
2. detailed and accurate docs for implementers to implement OAM on whatever runtime, and what's more important, with minimal effort. For example, again the `metadata` section, the message will be much clearer like "OAM metadata is a subset of K8s metadata (link) and here're the fields OAM cares".

I don't see any reason that we can't achieve this or have to mix them in one spec doc. Though I'd suggest we move this discussion out of this PR, it need to be fixed by following ups (and would be better to involve a tech writer) /cc @ryanzhang-oss